mutation_writer/shard_based_splitting_writer.cc

```diff
@@ -0,0 +1,130 @@
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mutation_writer/shard_based_splitting_writer.hh"
+
+#include <cinttypes>
+#include <boost/range/adaptor/transformed.hpp>
+#include <boost/range/algorithm/min_element.hpp>
+#include <seastar/core/shared_mutex.hh>
+
+#include "mutation_reader.hh"
+
+namespace mutation_writer {
+
+class shard_based_splitting_mutation_writer {
+    class shard_writer {
+        schema_ptr _schema;
+        queue_reader_handle _handle;
+        future<> _consume_fut;
+    private:
+        shard_writer(schema_ptr schema, std::pair<flat_mutation_reader, queue_reader_handle> queue_reader, reader_consumer& consumer)
+            : _schema(std::move(schema))
+            , _handle(std::move(queue_reader.second))
+            , _consume_fut(consumer(std::move(queue_reader.first))) {
+        }
+
+    public:
+        shard_writer(schema_ptr schema, reader_consumer& consumer)
+            : shard_writer(schema, make_queue_reader(schema), consumer) {
+        }
+        future<> consume(mutation_fragment mf) {
+            return _handle.push(std::move(mf));
+        }
+        future<> consume_end_of_stream() {
+            _handle.push_end_of_stream();
+            return std::move(_consume_fut);
+        }
+    };
+
+private:
+    schema_ptr _schema;
+    reader_consumer _consumer;
+    unsigned _current_shard;
+    std::vector<std::optional<shard_writer>> _shards;
+private:
+    future<> write_to_shard(mutation_fragment&& mf);
+public:
+    shard_based_splitting_mutation_writer(schema_ptr schema, reader_consumer consumer)
+        : _schema(std::move(schema))
+        , _consumer(std::move(consumer))
+        , _shards(smp::count)
+    {}
+
+    future<> consume(partition_start&& ps);
+    future<> consume(static_row&& sr);
+    future<> consume(clustering_row&& cr);
+    future<> consume(range_tombstone&& rt);
+    future<> consume(partition_end&& pe);
+
+    future<> consume_end_of_stream() {
+        return parallel_for_each(_shards, [] (std::optional<shard_writer>& shard) {
+            if (shard) {
+                return shard->consume_end_of_stream();
+            }
+            return make_ready_future<>();
+        });
+    }
+};
+
+future<> shard_based_splitting_mutation_writer::write_to_shard(mutation_fragment&& mf) {
+    auto& writer = *_shards[_current_shard];
```

Comment:
I am not a specialist in mutation fragment processing, but I don't nderstand what you mean.
I thought we were fed the fragments piecemeal at the resolution of a partition start, clustering row, etc...

How can I unconditionally process an entire partition ?

Reply:
Yes you are fed fragments one at a time. What I meant that when a new `partition_start` fragment comes in, instead of just setting `_current_shard`, then on each subsequent fragment doing a lookup for the appropriate reader, you could save a pointer to the writer instead, as you know its not going to change until the next `partition_start` fragment.

Reply:
But that buys us what? Unlike the timestamp splitter, this is an array, not a set.
The lookup is really just an arithmetic add.

Reply:
Yes but why even do it if not needed? Something being very cheap is not a reason to do it if there is really no reason to do it.

Reply:
Disagree.
I need a vector anyway to keep track of which writers exist and which don't.
Adding a pointer is more expensive in terms of memory, and really there isn't anything cheaper than an arithmetic add.

Reply:
Ok, I'm not going to make you benchmark 2 saved bytes vs. an additional add. :))

Reply:
Sorry, 4 saved bytes. This changes everyting.